### PR TITLE
added warning if mutable rev is used

### LINKED ISSUE #974: issue a warning if a mutable ref is used (`HEAD` / `stable` / `master` / etc.)
Perhaps nudging towards https://pre-commit.com/#using-the-latest-version-for-a-repository or `pre-commit autoupdate`

### PR DESCRIPTION
**Pull request by paulhfischer**
resolves #974

- only added a warning instead of raising an error because this might break multiple projects otherwise
- not sure if other mutable revs should also cause a warning, couldn't think of a way to detect them in general …
- the check could also be included in the `MigrateShaToRev`-class, but i think it's better to keep them separated
- test only checks for `HEAD`, this is sufficient imo, but i can also change it so it tests for all mutable revs defined in `{'HEAD', 'stable', 'master'}`

### CODE REVIEW

File: pre_commit/clientlib.py [hunk 1]
```diff
@@ -112,6 +112,24 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
 META = 'meta'
 
 
+class CheckMutableRev:
```
Comment (asottile) [inline pre_commit/clientlib.py:115]: can you use the pattern I adjusted your code to in the other PR?  I believe this should be a `Conditional` similar to the one in `MigrateShaToRev` (which will eventually be deleted)

File: pre_commit/clientlib.py [hunk 2]
```diff
@@ -112,6 +112,24 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+    key = 'rev'
+
+    def check(self, dct: Dict[str, Any]) -> None:
+        if any(
+            mutable in dct.get(self.key, '')
+            for mutable in {'HEAD', 'stable', 'master'}
```
Comment (asottile) [inline pre_commit/clientlib.py:121]: I'm not sure this will catch all the cases, notably it's missing `''`, `'main'` at the very least

I wonder if there's an easier way to catch these -- maybe match anything that doesn't look like a version?

File: pre_commit/clientlib.py [hunk 3]
```diff
@@ -112,6 +112,24 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+        ):
+            logger.warning(
+                f"The {self.key!r} field of repo {dct.get('repo')!r} "
+                f"shouldn't be mutable",
```
Comment (asottile) [inline pre_commit/clientlib.py:125]: the error message should link to the documentation

File: pre_commit/clientlib.py [hunk 4]
```diff
@@ -112,6 +112,24 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+class CheckMutableRev:
+    key = 'rev'
+
+    def check(self, dct: Dict[str, Any]) -> None:
+        if any(
+            mutable in dct.get(self.key, '')
+            for mutable in {'HEAD', 'stable', 'master'}
+        ):
+            logger.warning(
+                f"The {self.key!r} field of repo {dct.get('repo')!r} "
```
Comment (asottile) [inline pre_commit/clientlib.py:124]: repo is required by this point

File: pre_commit/clientlib.py [hunk 5]
```diff
@@ -112,6 +112,24 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+                f"shouldn't be mutable",
+            )
+
+    apply_default = cfgv.Required.apply_default
+
+    remove_default = cfgv.Required.remove_default
```
Comment (asottile) [inline pre_commit/clientlib.py:130]: `rev` isn't `Required`, for example `repo: local` / `repo: meta` don't use it

Comment (paulhfischer) [reply pre_commit/clientlib.py:130]: didn't think of that, changed it to `Conditional`

File: pre_commit/clientlib.py (hunk 1 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py:115]: i didn't change it to inherit from `cfgv.Conditional` but took a similar approach as `MigrateShaToRev`, as an inheritance would require passing in the arguments `key`, `check_fn`, `condition_key` and `condition_value` which should be hard coded in `_cond` in my opinion

File: pre_commit/clientlib.py (hunk 4 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py:124]: changed it to `dct['repo']` if that's what you referred to here

File: pre_commit/clientlib.py (hunk 2 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py:130]: alternative idea (not added yet):

```python
rev = dct.get(self.key, "")

if (
    "." not in rev and (  # not a version
        len(rev) != 40 or  # full sha
        len(rev) != 7 or  # short sha
        len(rev) != 8 or  # gitlab lists 8 char long sha in the commit-list
    ) or
    not re.match(r"^[a-zA-Z0-9]+$", rev)  # contains other chars (e.g. "_", "-", …)
)
# print warning
```

Comment (asottile) [reply pre_commit/clientlib.py:130]: Yeah that might be better -- I wouldn't validate length at all though -- you could have a 15 character "short sha" for example so that would lead to a false positive

It might be worth trying to implement some sort of heuristic and then searching github to find if things adhere to it?

File: pre_commit/clientlib.py [hunk 6]
```diff
@@ -112,6 +112,34 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+        ):
+            logger.warning(
+                f'The {self.key!r} field of repo {dct["repo"]!r} '
+                f"shouldn't be mutable. See https://pre-commit.com/#using-the-"
+                f'latest-version-for-a-repository for further details.',
+            )
+
+    apply_default = cfgv.Conditional.apply_default
+
+    remove_default = cfgv.Conditional.remove_default
```
Comment (asottile) [inline pre_commit/clientlib.py:140]: I'd still rather see this inherit from `Conditional` as I plan to delete `MigrateShaToRev` in pre-commit 3.x

File: pre_commit/clientlib.py (hunk 2 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py:130]: so i took a look at current usages of `rev` in `.pre-commit-config.yaml`-files (via github api [`https://api.github.com/search/code?q=filename:%2Epre-commit-config%2Eyaml&page={page}`]) and extracted all revs field. here's what i got so far:

- we can detect shas in general with `r"^[a-z0-9]+$"`, as they all only contain **lowercase** letters and numbers. this will result in false positives for branches, tho.
- version tags are all messed up. most of them follow the pattern `r"^v?([0-9]+\.)+[0-9]$"`, but:
  - some (e.g. ["20.8b1"](https://github.com/psf/black/releases/tag/20.8b1)) contain letters
  - some (e.g. ["4.3.21-2"](https://github.com/PyCQA/isort/releases/tag/4.3.21-2)) contain dashes
  thus, the only reasonable way to detect tags - in my opinion - is, to whitelist all strings that contain dots. (e.g contains `r"[^\.]\.[^\.]"`)
- negative examples i found are:
  - `"stable"`
  - `"master"`
  - `""`

thus, this is my suggestion (tested it against the data i got from the api calls):
```python
bad_revs = {
    "stable",
    "master",
    "HEAD",
}
rev = dct.get(self.key, "")

if "." not in rev and (
    not re.match(r"^[a-z0-9]+$", rev)
    or any(bad_rev in rev for bad_rev in bad_revs)
    or not rev
):
    print(f"Warning: {rev}")
```

File: pre_commit/clientlib.py (hunk 6 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py]: i had to change the inheritance to `ConditionalOptional`, due to the support of `sha` instead of `rev`:

```console
FAILED tests/commands/migrate_config_test.py::test_migrate_config_sha_to_rev - pre_commit.clientlib.InvalidConfigError:
```

also added a comment that the inheritance can (and should) be changed to `Conditional` if support for `sha` (e.g. deletion of `MigrateShaToRev`) is dropped.

File: tests/clientlib_test.py [hunk 7]
```diff
@@ -166,6 +166,54 @@ def test_validate_warn_on_unknown_keys_at_top_level(tmpdir, caplog):
...
+
+    expected = [
+        (
+            'pre_commit',
+            logging.WARNING,
+            "The 'rev' field of repo 'https://gitlab.com/pycqa/flake8' "
+            "shouldn't be mutable. See https://pre-commit.com/#using-the-"
+            'latest-version-for-a-repository for further details.',
+        ),
+    ] if error else []
```
Comment (asottile) [inline tests/clientlib_test.py:199]: please don't put logic in tests, this should be two separate tests, one demonstrating positive cases and one demonstrating negatives -- as it is right now you've introduced a boolean trap via parametrize (a common antipattern)

File: pre_commit/clientlib.py [hunk 8]
```diff
@@ -112,6 +113,32 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+class CheckBadRev(cfgv.ConditionalOptional):
+    def check(self, dct: Dict[str, Any]) -> None:
+        super().check(dct)
+
+        if self.key in dct.keys():
+            bad_revs = {
+                'stable',
+                'master',
+                'HEAD',
+            }
```
Comment (asottile) [inline pre_commit/clientlib.py:126]: this should be a module constant if we keep it, I don't think we should though

File: pre_commit/clientlib.py [hunk 9]
```diff
@@ -112,6 +113,32 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+            rev = dct[self.key]
+
+            if '.' not in rev and (
+                not re.match(r'^[a-z0-9]+$', rev) or
+                any(bad_rev in rev for bad_rev in bad_revs) or
+                not rev
```
Comment (asottile) [inline pre_commit/clientlib.py:132]: the empty string should warn too, it's (accidentally) an alias for `HEAD`

File: pre_commit/clientlib.py [hunk 10]
```diff
@@ -112,6 +113,32 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+            bad_revs = {
+                'stable',
+                'master',
+                'HEAD',
+            }
+            rev = dct[self.key]
+
+            if '.' not in rev and (
+                not re.match(r'^[a-z0-9]+$', rev) or
+                any(bad_rev in rev for bad_rev in bad_revs) or
```
Comment (asottile) [inline pre_commit/clientlib.py:131]: I think if you delete this line it'll still function the same -- also these should do exact matching, `'HEAD'` in `rev` is nonsense

File: pre_commit/clientlib.py [hunk 11]
```diff
@@ -112,6 +113,32 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
 META = 'meta'
 
 
+# should inherit from cfgv.Conditional if sha support is dropped
+class CheckBadRev(cfgv.ConditionalOptional):
```
Comment (asottile) [inline pre_commit/clientlib.py:117]: `WarnMutableRev` is probably a better name

File: pre_commit/clientlib.py [hunk 12]
```diff
@@ -112,6 +113,32 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+    def check(self, dct: Dict[str, Any]) -> None:
+        super().check(dct)
+
+        if self.key in dct.keys():
```
Comment (asottile) [inline pre_commit/clientlib.py:121]: `.keys()` should almost never be called -- there's only one valid use case that I've ever seen for it (set operations (`|` / `&`) against keys) and every other use can be replaced by removing `.keys()`, you want `if self.key in dct`

File: pre_commit/clientlib.py [hunk 13]
```diff
@@ -112,6 +113,32 @@ def validate_manifest_main(argv: Optional[Sequence[str]] = None) -> int:
...
+            bad_revs = {
+                'stable',
+                'master',
+                'HEAD',
+            }
+            rev = dct[self.key]
+
+            if '.' not in rev and (
+                not re.match(r'^[a-z0-9]+$', rev) or
```
Comment (asottile) [inline pre_commit/clientlib.py:130]: upper case shas are valid git revisions, this should be `a-f0-9A-F0-9`

File: pre_commit/clientlib.py (hunk 9 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py:132]: isn't `not rev` creating a warning for empty strings?

File: pre_commit/clientlib.py (hunk 13 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py:130]: good point! i forgot that letters are also restricted to `a-f` …

File: pre_commit/clientlib.py (hunk 9 above)
Comment (paulhfischer) [reply pre_commit/clientlib.py:126]: i also added a test case for an empty `rev` string

Comment (asottile) [reply pre_commit/clientlib.py:126]: yeah I think you can remove the `or not rev` -- `'.'` is not contained in `''` and `''` does not match your pattern

File: tests/clientlib_test.py [hunk 14]
```diff
@@ -166,6 +166,72 @@ def test_validate_warn_on_unknown_keys_at_top_level(tmpdir, caplog):
     ]
 
 
+@pytest.mark.parametrize(
+    ('rev'),
+    (
+        ('v0.12.4'),
```
[the remark this answers is not in the document]
Comment (paulhfischer) [reply tests/clientlib_test.py]: oh good to know! thanks for the note, i changed it to match the signature.

**Review by asottile — APPROVED** [review]
